### PR TITLE
First implementation of Soft Actor Critic for discrete actions

### LINKED ISSUE #184: Add SAC algorithm

### PR DESCRIPTION
**Pull request by Corentin-pro**
Fixes #184

### CODE REVIEW

File: ai/src/algorithms/sacd/agent.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+            with tf.GradientTape() as alpha_tape:
+                alpha_loss = tf.reduce_mean(-1 * self.log_alpha * (neg_entropy + self.target_entropy))
+        self._alpha_optimizer.apply_gradients(
+            zip(alpha_tape.gradient(alpha_loss, [self.log_alpha]), [self.log_alpha]))
+        self.alpha = tf.exp(self.log_alpha)
+
+        self._copy_target_models()
+
+
+class SACDAgent(SpiceAIAgent):
```
Comment (phillipleblanc) [inline ai/src/algorithms/sacd/agent.py]:
```suggestion
class SoftActorCriticDiscreteAgent(SpiceAIAgent):
```

I would suggest making this explicit - someone reading the code in the factory method won't know what SACD stands for.

Comment (Corentin-pro) [reply ai/src/algorithms/sacd/agent.py]: Yes good idea, I was too used to read SAC and just named it like this

File: ai/src/algorithms/sacd/sacd.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+    "algorithm_id": "sacd",
+    "name": "Soft Actor Critic Discrete",
+    "docs_link": "https://docs.spiceai.org/deep-learning-ai/sac/"
```
Comment (phillipleblanc) [inline ai/src/algorithms/sacd/sacd.json:4]: Are you planning to make this docs change today? Otherwise I'll create an issue to track it.

**Review by phillipleblanc — COMMENTED** [review]
Looks good! We should also add E2E test coverage of this new algorithm. I'm not sure if it makes sense to test the entire matrix of [metal, docker] x [dql, vpg, sacd] - but we can start with that for now and I can add an item to filter this down later. Once we make the soft actor critic the default, I expect we can just run a single test for both dql/vpg and only run the full matrix for sacd

File: ai/src/algorithms/sacd/sacd.json (hunk 2 above)
Comment (Corentin-pro) [reply ai/src/algorithms/sacd/sacd.json:4]: Let's make an issue indeed, this is probably be made latter

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**